src/Template.php

```diff
@@ -18,191 +18,131 @@ class Template implements \ArrayAccess
-        }
+        if ($ref !== null) {
+            if (!isset($this->tagsIndex[$tag][$ref])) {
+                throw $this->exceptionAddMoreInfo(
+                    (new Exception('Tag not found in template'))
+                        ->addMoreInfo('tag', $tag . '#' . $ref)
+                );
+            }
 
-        $a = explode('#', $tag);
-        $tag = array_shift($a);
-        //$ref = array_shift($a); // unused
-        if (!isset($this->tags[$tag])) {
-            throw $this->exception('Tag not found in Template')
-                ->addMoreInfo('tag', $tag)
-                ->addMoreInfo('tags', implode(', ', array_keys($this->tags)));
+            $path = $this->tagsIndex[$tag][$ref];
+        } else {
+            if ($tag === self::TOP_TAG) {
+                return $this->template;
+            }
+
+            $tag = explode('#', $tag, 2)[0];
+            if (!isset($this->tagsIndex[$tag])) {
+                throw $this->exceptionAddMoreInfo(
+                    (new Exception('Tag not found in template'))
+                        ->addMoreInfo('tag', $tag)
+                );
+            }
+
+            $path = reset($this->tagsIndex[$tag]);
         }
 
-        // return first array element only
-        reset($this->tags[$tag]);
-        $key = key($this->tags[$tag]) !== null ? key($this->tags[$tag]) : null;
+        $vRef = &$this->template;
+        foreach ($path as $k) {
+            $vRef = &$vRef[$k];
+        }
 
-        return $this->tags[$tag][$key];
+        return $vRef;
     }
 
-    /**
-     * For methods which execute action on several tags, this method
-     * will return array of templates. You can then iterate
-     * through the array and update all the template values.
-     *
-     * {greeting}hello{/},  {greeting}world{/}
-     *
-     * calling $template =& getTagRefList('greeting') will point
-     * $template towards array(&array('hello'),&array('world'));
-     *
-     * If $tag is specified as an array, then $template will
-     * contain all occurrences of all tags from the array.
-     *
-     * @param string|array $tag
-     *
-     * @return array of references to template tags
-     */
-    public function getTagRefList($tag)
+    protected function getTagRefs(string $tag): array
     {
-        if (is_array($tag)) {
-            $res = [];
-            foreach ($tag as $t) {
-                $list = $this->getTagRefList($t);
-                foreach ($list as &$tpl) {
-                    $res[] = &$tpl;
-                }
-            }
-
-            return $res;
-        }
-
-        if ($this->isTopTag($tag)) {
+        if ($tag === self::TOP_TAG) {
             return [&$this->template];
         }
 
-        $a = explode('#', $tag);
-        $tag = array_shift($a);
-        $ref = array_shift($a);
-        if (!$ref) {
-            if (!isset($this->tags[$tag])) {
-                throw $this->exception('Tag not found in Template')
+        $tag = explode('#', $tag, 2)[0];
```

Comment:
This code below slightly duplicate code in `getTagRef` method.

Reply:
Yes, but we need to access `tagsIndex` first. Later, we should probably store values of template structure in 1D array to fix the rest performance issues. (I expect like +200% faster code)